packages/langium/src/grammar/lsp/grammar-call-hierarchy.ts

```diff
@@ -0,0 +1,106 @@
+ * Copyright 2021 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { CallHierarchyIncomingCall, CallHierarchyOutgoingCall, Range, SymbolKind } from 'vscode-languageserver';
+import { DefaultCallHierarchyProvider } from '../../lsp/call-hierarchy-provider';
+import { AstNode, CstNode } from '../../syntax-tree';
+import { getContainerOfType, getDocument, streamAllContents } from '../../utils/ast-util';
+import { findLeafNodeAtOffset } from '../../utils/cst-util';
+import { ReferenceDescription } from '../../workspace/ast-descriptions';
+import { isParserRule, isRuleCall, RuleCall } from '../generated/ast';
+
+export class LangiumGrammarCallHierarchyProvider extends DefaultCallHierarchyProvider {
+    protected getIncomingCalls(references: ReferenceDescription[]): CallHierarchyIncomingCall[] | null {
+        // This map is used to group incoming calls to avoid duplicates.
+        const uniqueRules = new Map<string, { nameNode: CstNode, targetNodes: CstNode[], docUri: string }>();
+        references.forEach(ref => {
+            const doc = this.documents.getOrCreateDocument(ref.sourceUri);
+            const rootNode = doc.parseResult.value;
+            if (!rootNode.$cstNode) {
+                return;
+            }
+            const targetNode = findLeafNodeAtOffset(rootNode.$cstNode, ref.segment.offset);
+            if (!targetNode) {
+                return;
+            }
+            const parserRule = getContainerOfType(targetNode.element, isParserRule);
+            if (!parserRule) {
+                return;
+            }
+            const nameNode = this.nameProvider.getNameNode(parserRule);
+            if (!nameNode) {
+                return;
+            }
+            const refDocUri = ref.sourceUri.toString();
+            const ruleId = refDocUri + '@' + nameNode.text;
+            uniqueRules.has(ruleId) ?
+                uniqueRules.set(ruleId, { nameNode, targetNodes: [...uniqueRules.get(ruleId)!.targetNodes, targetNode], docUri: refDocUri })
+                : uniqueRules.set(ruleId, { nameNode, targetNodes: [targetNode], docUri: refDocUri });
+        });
+        if (uniqueRules.size === 0) {
+            return null;
+        }
+        const incomingCalls: CallHierarchyIncomingCall[] = [];
+        uniqueRules.forEach(nodes => {
+            incomingCalls.push({
+                from: {
+                    kind: SymbolKind.Method,
+                    name: nodes.nameNode.text,
+                    range: nodes.nameNode.range,
+                    selectionRange: nodes.nameNode.range,
+                    uri: nodes.docUri
+                },
+                fromRanges: nodes.targetNodes.map(node => node.range)
+            });
+        });
+        return incomingCalls;
```

Comment:
Minor: This can be mapped to a single `map` function call. (instead of `forEach` and `push`)

Reply:
I couldn't find a way to use a `map` function on a `Map` object, so I changed the map to an array and introduced an interface to help with the structure of the objects in the array

Reply:
You can do `Array.from(uniqueRules.values()).map(...)` instead.

Reply:
Went an other way, but can change it back if it's better